File: shared/src/api/sourcing.ts
// Requisitions, RFQs, supplier quotes and the award (blueprint B5, B5.1).
//
// # The comparison carries no winner
//
// `Comparison.lowest_quote_id` names the cheapest quote that could actually be
// accepted today, and that is all it is: a convenience for the eye. B5.1 wants
// a person to choose and to say why, because lead time and payment terms
// routinely outweigh price — so nothing here ranks, recommends, or pre-selects.

import type { Client } from './client';

export type RequisitionStatus =
  | 'draft'
  | 'submitted'
  | 'approved'
  | 'rejected'
  | 'sourcing'
  | 'ordered'
  | 'cancelled';

export interface RequisitionLine {
  id: string;
  line_no: number;
  variant_id?: string;
  sku?: string;
  description: string;
  qty_requested: string;
  note?: string;
}

export interface Requisition {
  id: string;
  requisition_no: string;
  status: RequisitionStatus;
  needed_by?: string;
  justification?: string;

  store_id?: string;
  store_name?: string;
  warehouse_id?: string;

  requested_by?: string;
  requested_at: string;
  decided_by?: string;
  decided_at?: string;
  decision_note?: string;

  lines?: RequisitionLine[];
}

export type RFQStatus =
  | 'draft'
  | 'issued'
  | 'comparing'
  | 'awarded'
  | 'cancelled';

export interface RFQLine {
  id: string;
  line_no: number;
  variant_id: string;
  sku?: string;
  description: string;
  qty: string;
}

export interface RFQInvited {
  supplier_id: string;
  supplier_name: string;
  invited_at: string;
  /** Asked and answered. Distinct from `declined_at`, because "asked and never
   *  replied" is information a buyer needs about a supplier. */
  quoted: boolean;
  declined_at?: string;
  decline_reason?: string;
}

export interface RFQ {
  id: string;
  rfq_number: string;
  status: RFQStatus;
  requisition_id?: string;
  warehouse_id: string;
  closes_on?: string;
  currency: string;
  notes?: string;

  awarded_quote_id?: string;
  award_reason?: string;
  awarded_by?: string;
  awarded_at?: string;
  cancel_reason?: string;

  created_at: string;
  issued_at?: string;

  lines?: RFQLine[];
  invited?: RFQInvited[];
}

export interface QuoteLine {
  id: string;
  rfq_line_id: string;
  line_no: number;
  variant_id: string;
  sku?: string;
  description: string;
  qty: string;
  unit_cost: string;
  tax_treatment: string;
  net_amount: string;
  tax_amount: string;
  gross_amount: string;
}

export interface Quote {
  id: string;
  rfq_id: string;
  supplier_id: string;
  supplier_name: string;
  quote_number?: string;
  revision: number;
  status: 'received' | 'superseded' | 'awarded' | 'rejected';

  received_on: string;
  valid_until?: string;
  /** Derived by comparing the validity date to today, never stored: a quote
   *  does not become a different document at midnight. */
  expired: boolean;

  currency: string;
  lead_time_days?: number;
  payment_terms_days?: number;
  quality_note?: string;

  subtotal_net: string;
  tax_total: string;
  total_inclusive: string;
  notes?: string;

  lines?: QuoteLine[];
}

export interface Comparison {
  rfq: RFQ;
  quotes: Quote[];
  lowest_quote_id?: string;
}

const scoped = (path: string, companyId: string) =>
  `${path}${path.includes('?') ? '&' : '?'}company_id=${companyId}`;

const query = (q: Record<string, string | undefined>) => {
  const parts = Object.entries(q)
    .filter(([, v]) => v !== undefined && v !== '')
    .map(([k, v]) => `${k}=${encodeURIComponent(String(v))}`);
  return parts.length ? '?' + parts.join('&') : '';
};

// --- Requisitions --------------------------------------------------------

export function listRequisitions(
  client: Client,
  companyId: string,
  q: { status?: string } = {},
): Promise<{ requisitions: Requisition[] }> {
  return client.send<{ requisitions: Requisition[] }>(
    'GET',
    scoped('/api/v1/purchasing/requisitions' + query(q), companyId),
  );
}

export function readRequisition(
  client: Client,
  companyId: string,
  id: string,
): Promise<Requisition> {
  return client.send<Requisition>(
    'GET',
    scoped(`/api/v1/purchasing/requisitions/${id}`, companyId),
  );
}

export function raiseRequisition(
  client: Client,
  companyId: string,
  body: {
    store_id?: string;
    warehouse_id?: string;
    needed_by?: string;
    justification?: string;
    lines: Array<{
      variant_id?: string;
      description?: string;
      qty: string;
      note?: string;
    }>;
  },
): Promise<Requisition> {
  return client.send<Requisition>(
    'POST',
    scoped('/api/v1/purchasing/requisitions', companyId),
    body,
  );
}

export function decideRequisition(
  client: Client,
  companyId: string,
  id: string,
  approve: boolean,
  note?: string,
): Promise<Requisition> {
  return client.send<Requisition>(
    'POST',
    scoped(`/api/v1/purchasing/requisitions/${id}/decision`, companyId),
    { approve, note },
  );
}

// --- RFQ -----------------------------------------------------------------

export function listRFQs(
  client: Client,
  companyId: string,
  q: { status?: string } = {},
): Promise<{ rfqs: RFQ[] }> {
  return client.send<{ rfqs: RFQ[] }>(
    'GET',
    scoped('/api/v1/purchasing/rfqs' + query(q), companyId),
  );
}

export function raiseRFQ(
  client: Client,
  companyId: string,
  body: {
    requisition_id?: string;
    warehouse_id: string;
    closes_on?: string;
    notes?: string;
    supplier_ids: string[];
    lines: Array<{ variant_id: string; description?: string; qty: string }>;
  },
): Promise<RFQ> {
  return client.send<RFQ>(
    'POST',
    scoped('/api/v1/purchasing/rfqs', companyId),
    body,
  );
}

export function compareQuotes(
  client: Client,
  companyId: string,
  rfqId: string,
): Promise<Comparison> {
  return client.send<Comparison>(
    'GET',
    scoped(`/api/v1/purchasing/rfqs/${rfqId}/comparison`, companyId),
  );
}

export function recordQuote(
  client: Client,
  companyId: string,
  rfqId: string,
  body: {
    supplier_id: string;
    quote_number?: string;
    received_on?: string;
    valid_until?: string;
    lead_time_days?: number;
    payment_terms_days?: number;
    quality_note?: string;
    notes?: string;
    lines: Array<{
      rfq_line_id: string;
      qty: string;
      unit_cost: string;
      tax_treatment?: string;
      tax_rate?: string;
      note?: string;
    }>;
  },
): Promise<Quote> {
  return client.send<Quote>(
    'POST',
    scoped(`/api/v1/purchasing/rfqs/${rfqId}/quotes`, companyId),
    body,
  );
}

export function declineToQuote(
  client: Client,
  companyId: string,
  rfqId: string,
  supplierId: string,
  reason?: string,
): Promise<void> {
  return client.send<void>(
    'POST',
    scoped(`/api/v1/purchasing/rfqs/${rfqId}/declines`, companyId),
    { supplier_id: supplierId, reason },
  );
}

/** Picks the winner and raises the purchase order. The reason is mandatory:
 *  it is the record an audit reads, and cheapest does not always win. */
export function awardQuote(
  client: Client,
  companyId: string,
  rfqId: string,
  quoteId: string,
  reason: string,
): Promise<{ rfq: RFQ; order: { id: string; po_number: string } }> {
  return client.send(
    'POST',
    scoped(`/api/v1/purchasing/rfqs/${rfqId}/award`, companyId),
    { quote_id: quoteId, reason },
  );
}

export function cancelRFQ(
  client: Client,
  companyId: string,
  rfqId: string,
  reason: string,
): Promise<void> {
  return client.send<void>(
    'POST',
    scoped(`/api/v1/purchasing/rfqs/${rfqId}/cancel`, companyId),
    { reason },
  );
}

/** B5.1's archive: what this supplier has quoted before, won or lost. */
export function supplierQuoteHistory(
  client: Client,
  companyId: string,
  supplierId: string,
): Promise<{ quotes: Quote[] }> {
  return client.send<{ quotes: Quote[] }>(
    'GET',
    scoped(`/api/v1/purchasing/suppliers/${supplierId}/quotes`, companyId),
  );
}

File: backend/internal/purchasing/sourcing.go
// Sourcing: requisition, RFQ, supplier quotes and the award (blueprint B5,
// B5.1).
//
// This is the front half of the purchasing chain, and it ends where the back
// half begins: an award produces a purchase order through CreateOrder, the same
// function a buyer raising an order by hand reaches. There is deliberately no
// second path that writes a purchase_order row, because a PO born of a quote
// and a PO typed in by hand must be the same document in every respect that
// matters downstream — receiving, three-way matching and payment all read one
// shape.
//
// # What is compared, and what is not
//
// B5.1 wants price, total, VAT, lead time, payment terms and quality notes side
// by side. It does NOT want the system to pick a winner. Cheapest is often the
// wrong answer — a supplier who delivers in three days against one who takes
// three weeks, or one who bills on 60 days against cash on delivery — so the
// comparison presents the axes and a person decides, on the record, with a
// reason. Compare() therefore computes and ranks nothing beyond arithmetic the
// quotes already state.
package purchasing

import (
	"context"
	"errors"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/shopspring/decimal"

	"github.com/mahedi-emon/rawsyst-pos/backend/internal/platform/db"
	"github.com/mahedi-emon/rawsyst-pos/backend/internal/platform/errs"
)

// --- Requisitions --------------------------------------------------------

// NewRequisition is somebody asking for stock.
//
// No prices. B5's requisition is "any authorized staff can request stock"; the
// person who notices an empty shelf is not the person who negotiates cost, and
// asking them for one produces either a guess or an abandoned request.
type NewRequisition struct {
	StoreID     *uuid.UUID
	WarehouseID *uuid.UUID
	NeededBy    *time.Time
	Why         string
	Lines       []NewRequisitionLine
}

// NewRequisitionLine is one thing wanted.
//
// VariantID is optional: a shop that needs something it has never bought
// describes it in words, and the buyer turns that into a catalogue entry when
// they source it. Requiring a variant would mean nobody could ask for anything
// new.
type NewRequisitionLine struct {
	VariantID   *uuid.UUID
	Description string
	Qty         decimal.Decimal
	Note        string
}

// Requisition is a request as it is read back.
type Requisition struct {
	ID       uuid.UUID `json:"id"`
	Number   string    `json:"requisition_no"`
	Status   string    `json:"status"`
	NeededBy string    `json:"needed_by,omitempty"`
	Why      string    `json:"justification,omitempty"`

	StoreID     *uuid.UUID `json:"store_id,omitempty"`
	StoreName   string     `json:"store_name,omitempty"`
	WarehouseID *uuid.UUID `json:"warehouse_id,omitempty"`

	RequestedBy  string            `json:"requested_by,omitempty"`
	RequestedAt  string            `json:"requested_at"`
	DecidedBy    string            `json:"decided_by,omitempty"`
	DecidedAt    string            `json:"decided_at,omitempty"`
	DecisionNote string            `json:"decision_note,omitempty"`
	Lines        []RequisitionLine `json:"lines,omitempty"`
}

// RequisitionLine is one line as read back.
type RequisitionLine struct {
	ID          uuid.UUID  `json:"id"`
	LineNo      int        `json:"line_no"`
	VariantID   *uuid.UUID `json:"variant_id,omitempty"`
	SKU         string     `json:"sku,omitempty"`
	Description string     `json:"description"`
	Qty         string     `json:"qty_requested"`
	Note        string     `json:"note,omitempty"`
}

// RaiseRequisition records a request for stock.
//
// It is created 'submitted' rather than 'draft'. A draft that nobody can see is
// a request that never reaches an approver, and the shelf stays empty while the
// requester believes they have asked. Editing before submitting is a refinement
// worth having only once somebody complains they cannot.
func (s *Service) RaiseRequisition(
	ctx context.Context, scope Scope, in NewRequisition,
) (Requisition, error) {
	if len(in.Lines) == 0 {
		return Requisition{}, errs.New(errs.CodeInvalidInput,
			"A request needs at least one line.")
	}
	for i, l := range in.Lines {
		if !l.Qty.IsPositive() {
			return Requisition{}, errs.Newf(errs.CodeInvalidInput,
				"Line %d has no quantity.", i+1)
		}
		if l.VariantID == nil && strings.TrimSpace(l.Description) == "" {
			return Requisition{}, errs.Newf(errs.CodeInvalidInput,
				"Line %d names neither a product nor what is wanted.", i+1)
		}
	}

	var out Requisition
	err := s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		if in.StoreID != nil {
			if e := checkBelongs(ctx, tx, "store", *in.StoreID, scope.CompanyID,
				"That store was not found."); e != nil {
				return e
			}
		}
		if in.WarehouseID != nil {
			if e := checkBelongs(ctx, tx, "warehouse", *in.WarehouseID,
				scope.CompanyID, "That warehouse was not found."); e != nil {
				return e
			}
		}

		number, e := claimNumber(ctx, tx, scope.CompanyID, "requisition", "REQ")
		if e != nil {
			return e
		}

		var id uuid.UUID
		if e := tx.QueryRow(ctx, `
			INSERT INTO purchase_requisition
			  (tenant_id, company_id, requisition_no, store_id, warehouse_id,
			   status, needed_by, justification, requested_by)
			VALUES ($1,$2,$3,$4,$5,'submitted',$6,$7,$8) RETURNING id`,
			scope.TenantID, scope.CompanyID, number, in.StoreID, in.WarehouseID,
			in.NeededBy, nullText(in.Why), scope.UserID,
		).Scan(&id); e != nil {
			return e
		}

		for i, l := range in.Lines {
			desc := strings.TrimSpace(l.Description)
			if desc == "" && l.VariantID != nil {
				// Fall back to the catalogue's own words, so a line always
				// says what it is when it is read back months later.
				_ = tx.QueryRow(ctx,
					`SELECT coalesce(v.sku, '') FROM variant v WHERE v.id = $1`,
					*l.VariantID).Scan(&desc)
			}
			if _, e := tx.Exec(ctx, `
				INSERT INTO requisition_line
				  (tenant_id, requisition_id, line_no, variant_id, description,
				   qty_requested, note)
				VALUES ($1,$2,$3,$4,$5,$6,$7)`,
				scope.TenantID, id, i+1, l.VariantID, desc, l.Qty,
				nullText(l.Note)); e != nil {
				return e
			}
		}

		read, e := s.readRequisition(ctx, tx, scope.CompanyID, id)
		out = read
		return e
	})
	if err != nil {
		return Requisition{}, db.Translate(err, "")
	}
	return out, nil
}

// DecideRequisition approves or rejects a request.
//
// A rejection must say why. The requester has to be able to act on the answer —
// order less, order later, or stop asking — and "rejected" alone tells them
// none of those.
func (s *Service) DecideRequisition(
	ctx context.Context, scope Scope, id uuid.UUID, approve bool, note string,
) (Requisition, error) {
	if !approve && strings.TrimSpace(note) == "" {
		return Requisition{}, errs.Validation(
			"Say why the request is being turned down.").
			WithField("decision_note",
				"The person who asked needs to know what to do instead.")
	}

	status := "approved"
	if !approve {
		status = "rejected"
	}

	var out Requisition
	err := s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		tag, e := tx.Exec(ctx, `
			UPDATE purchase_requisition
			SET status = $3, decided_by = $4, decided_at = now(),
			    decision_note = $5
			WHERE id = $1 AND company_id = $2 AND status = 'submitted'`,
			id, scope.CompanyID, status, scope.UserID, nullText(note))
		if e != nil {
			return e
		}
		if tag.RowsAffected() == 0 {
			return requisitionNotDecidable(ctx, tx, scope.CompanyID, id)
		}

		read, e := s.readRequisition(ctx, tx, scope.CompanyID, id)
		out = read
		return e
	})
	if err != nil {
		return Requisition{}, db.Translate(err, "")
	}
	return out, nil
}

// requisitionNotDecidable says why an update matched nothing. Distinguishing
// "gone" from "already decided" is worth the second read: they lead an approver
// to do very different things next.
func requisitionNotDecidable(
	ctx context.Context, tx pgx.Tx, companyID, id uuid.UUID,
) error {
	var status string
	e := tx.QueryRow(ctx,
		`SELECT status FROM purchase_requisition WHERE id = $1 AND company_id = $2`,
		id, companyID).Scan(&status)
	if errors.Is(e, pgx.ErrNoRows) {
		return errs.New(errs.CodeNotFound, "That request was not found.")
	}
	if e != nil {
		return e
	}
	return errs.Newf(errs.CodeConflict,
		"That request is %s, so it cannot be decided again.", status)
}

// ListRequisitions returns requests, newest first.
func (s *Service) ListRequisitions(
	ctx context.Context, scope Scope, status string,
) ([]Requisition, error) {
	out := []Requisition{}
	err := s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		rows, e := tx.Query(ctx, `
			SELECT r.id, r.requisition_no, r.status, r.needed_by,
			       coalesce(r.justification, ''), r.store_id,
			       coalesce(st.name, ''), r.warehouse_id,
			       coalesce(u.full_name, ''), r.requested_at,
			       coalesce(d.full_name, ''), r.decided_at,
			       coalesce(r.decision_note, '')
			FROM purchase_requisition r
			LEFT JOIN store st   ON st.id = r.store_id
			LEFT JOIN app_user u ON u.id = r.requested_by
			LEFT JOIN app_user d ON d.id = r.decided_by
			WHERE r.company_id = $1 AND ($2 = '' OR r.status = $2)
			ORDER BY r.requested_at DESC
			LIMIT 500`, scope.CompanyID, status)
		if e != nil {
			return e
		}
		defer rows.Close()

		for rows.Next() {
			r, e := scanRequisition(rows)
			if e != nil {
				return e
			}
			out = append(out, r)
		}
		return rows.Err()
	})
	if err != nil {
		return nil, db.Translate(err, "")
	}
	return out, nil
}

// ReadRequisition returns one request with its lines.
func (s *Service) ReadRequisition(
	ctx context.Context, scope Scope, id uuid.UUID,
) (Requisition, error) {
	var out Requisition
	err := s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		read, e := s.readRequisition(ctx, tx, scope.CompanyID, id)
		out = read
		return e
	})
	if err != nil {
		return Requisition{}, db.Translate(err, "")
	}
	return out, nil
}

func (s *Service) readRequisition(
	ctx context.Context, tx pgx.Tx, companyID, id uuid.UUID,
) (Requisition, error) {
	row := tx.QueryRow(ctx, `
		SELECT r.id, r.requisition_no, r.status, r.needed_by,
		       coalesce(r.justification, ''), r.store_id,
		       coalesce(st.name, ''), r.warehouse_id,
		       coalesce(u.full_name, ''), r.requested_at,
		       coalesce(d.full_name, ''), r.decided_at,
		       coalesce(r.decision_note, '')
		FROM purchase_requisition r
		LEFT JOIN store st   ON st.id = r.store_id
		LEFT JOIN app_user u ON u.id = r.requested_by
		LEFT JOIN app_user d ON d.id = r.decided_by
		WHERE r.id = $1 AND r.company_id = $2`, id, companyID)

	out, err := scanRequisition(row)
	if errors.Is(err, pgx.ErrNoRows) {
		return Requisition{}, errs.New(errs.CodeNotFound,
			"That request was not found.")
	}
	if err != nil {
		return Requisition{}, err
	}

	rows, err := tx.Query(ctx, `
		SELECT l.id, l.line_no, l.variant_id, coalesce(v.sku, ''),
		       l.description, l.qty_requested, coalesce(l.note, '')
		FROM requisition_line l
		LEFT JOIN variant v ON v.id = l.variant_id
		WHERE l.requisition_id = $1
		ORDER BY l.line_no`, id)
	if err != nil {
		return Requisition{}, err
	}
	defer rows.Close()

	for rows.Next() {
		var l RequisitionLine
		var qty decimal.Decimal
		if e := rows.Scan(&l.ID, &l.LineNo, &l.VariantID, &l.SKU,
			&l.Description, &qty, &l.Note); e != nil {
			return Requisition{}, e
		}
		l.Qty = qty.String()
		out.Lines = append(out.Lines, l)
	}
	return out, rows.Err()
}

// rowScanner is satisfied by both pgx.Row and pgx.Rows, so one scan function
// serves the single read and the list.
type rowScanner interface {
	Scan(dest ...any) error
}

func scanRequisition(row rowScanner) (Requisition, error) {
	var r Requisition
	var neededBy, decidedAt *time.Time
	var requestedAt time.Time
	if err := row.Scan(&r.ID, &r.Number, &r.Status, &neededBy, &r.Why,
		&r.StoreID, &r.StoreName, &r.WarehouseID, &r.RequestedBy,
		&requestedAt, &r.DecidedBy, &decidedAt, &r.DecisionNote); err != nil {
		return Requisition{}, err
	}
	r.RequestedAt = requestedAt.UTC().Format(time.RFC3339)
	if neededBy != nil {
		r.NeededBy = neededBy.Format("2006-01-02")
	}
	if decidedAt != nil {
		r.DecidedAt = decidedAt.UTC().Format(time.RFC3339)
	}
	r.Lines = []RequisitionLine{}
	return r, nil
}
